### ISSUE TITLE
TF-2.0: Keras model save load memory leakage

### ISSUE THREAD

**Issue by julyrashchenko** [tensorflow/tensorflow#32234]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.3 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): tensorflow-gpu==2.0.0-beta1
- Python version: Python 3.6.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CUDA 10.0.130/ cuDNN 7.6
- GPU model and memory: GeForce GTX 1050 Ti, 4036 MiB

**Describe the current behavior**
During saving and loading model I face memory leakage. Eventually it crashes with `OSError: [Errno 12] Cannot allocate memory`. During real training it happens quite fast.

**Describe the expected behavior**
I expect that memory should be cleaned.

**Code to reproduce the issue**
```
import tensorflow as tf
import numpy as np
from tqdm import tqdm
from memory_profiler import profile

data_array = np.random.random_sample((1, 1024))
tf_array = tf.constant(data_array, dtype=tf.float32)

input = tf.keras.Input((1, 1024))
hidden_layer = tf.keras.layers.Dense(1024)(input)
output = tf.keras.layers.Dense(1)(hidden_layer)
model = tf.keras.Model(inputs=[input], outputs=[output])

pred = model([tf_array])
print(pred)


@profile
def func():
    export_path = "temp_export"
    tf.saved_model.save(model, export_path)
    imported = tf.saved_model.load(export_path)


for i in tqdm(range(1000000), total=1000000):
    func()
```

**Other info / logs**
Profiler logs (2 steps)
![Снимок экрана от 2019-09-05 15-14-15](https://user-images.githubusercontent.com/54936869/64340897-de95bf80-cfef-11e9-8069-4a8c532cd059.png)

**Comment by gowthamkpr** [comment]
@julyrashchenko Can you please provide the github gist of the issue. Thanks!

**Comment by julyrashchenko** [comment]
@gowthamkpr 
I managed to reproduce it on Google colab, here is the [gist](https://colab.research.google.com/drive/1If-E-UR23-ytIIIYctXScoJOMxZp42Xc). Is it okay? Also I attached output of my code running. You can see, that at the end there is memory leakage of about 4 Gb.
![image](https://user-images.githubusercontent.com/54936869/64475250-01a7a700-d189-11e9-978f-a225a1d4db1e.png)

**Comment by jvishnuvardhan** [comment]
@julyrashchenko Can you try `!pip install -q tf-nightly-gpu-2.0-preview` and let us know. With `tf-nightly` my output looks little different. ~100 MB per 100 iterations doesn't look bad i guess. @aaroey will know better. You could also delete those `imported` models. Thanks!

```
0 : free memory 11185.9296875 Mb
100 : free memory 11073.01171875 Mb
200 : free memory 10961.89453125 Mb
300 : free memory 10849.40625 Mb
400 : free memory 10738.34765625 Mb
500 : free memory 10624.68359375 Mb
600 : free memory 10512.73828125 Mb
700 : free memory 10398.15234375 Mb
800 : free memory 10285.1875 Mb
900 : free memory 10172.0859375 Mb
1000 : free memory 10057.1875 Mb
1100 : free memory 9946.0859375 Mb
1200 : free memory 9834.51171875 Mb
1300 : free memory 9723.73046875 Mb
```

**Comment by julyrashchenko** [comment]
Yes, I also see that with such settings (tf-nightly-gpu-2.0-preview) memory leak decreased to about 100 Mb per 100 iterations. Usage `del imported` didn't changed situation (~100 Mb per 100 steps, too). But I  think it is still a huge problem because leak increases with complicating the model. E.g., when I add a couple of dense layers, leak is about 200 Mb per 100 steps.
![Снимок экрана от 2019-09-30 14-18-32](https://user-images.githubusercontent.com/54936869/65874031-6ec7ea00-e38d-11e9-867b-d6e7b3dec7d1.png)

**Comment by ylipacbio** [comment]
Let me add to this thread what I saw w.r.t. tensorflow 2.0 memory leak in LoadSavedModel.

-  System info

```model name      : Intel(R) Xeon(R) CPU           X5670  @ 2.93GHz
LSB Version:    :core-4.1-amd64:core-4.1-noarch
Distributor ID: CentOS
Description:    CentOS Linux release 7.4.1708 (Core)
Release:        7.4.1708
gcc: (crosstool-NG 1.24.0) 8.3.0
```
- Tensorflow versions
I detected this memory leak in both release `2.0.0` and `r2.1` branch.

- Background 
I trained a tensorflow model in python and then loaded this model in C++ to do inference, then I detected memory leak using `-fsanitize leak`.

To reproduce this memory leak, any arbitrary simple TF model can be used. 

- Step 1: Save an arbitrary trained model in Python
Python code that I used to save a TF model to a directory.
```
    evl = model.evaluate(x_train, y_train)
    tf.saved_model.save(model, out_dir)
```
 , and under output directory 
```
ls out_dir
saved_model.pb assets/ variables/
```

- Step2: main.cpp in C++
Below is the C++ main.cpp which simply loads saved model from a directory, check status and then exit.  Please replace TRAINED_MODEL_PATH with your own model path before compile.
```
#include <tensorflow/cc/saved_model/loader.h>         // SavedModelBundle
#include <tensorflow/cc/saved_model/tag_constants.h>  // kSavedModelTagServe
#include <tensorflow/core/framework/tensor.h>         // Tensor
#include <tensorflow/core/platform/init_main.h>       // Init_Main
#include <iostream>
#include <stdexcept>

int main(int argc, char** argv)
{
    tensorflow::SavedModelBundle bundle;
    tensorflow::SessionOptions sessionOptions;
    tensorflow::RunOptions runOptions;
    tensorflow::Status load_graph_status = LoadSavedModel(
        sessionOptions, runOptions, "TRAINED_MODEL_PATH",
        {tensorflow::kSavedModelTagServe}, &bundle);
    // Check load was successful
    if (!load_graph_status.ok()) {
        std::cout << "Error!" << std::endl;
        return -1;
    }
    return 0;
}
```
Bash cmd to compile `main.cpp` with ASAN, USAN, and leak detection.
```
g++ -fsanitize=address -fsanitize=leak -fsanitize=undefined main.cpp \
    -Ipath_to_tensorflow_include \
    -ltensorflow_cc -ltensorflow_framework -lrt \
    -Lpath_to_tensorflow_lib \
    -o a.out
```
If you have installed `libtensorflow` and configed using `pkg_config`, you may alternative use:
```
g++ -fsanitize=address -fsanitize=leak -fsanitize=undefined main.cpp $(pkg-config --cflags tensorflow) $(pkg-config --libs tensorflow) -o a.out
```

Bash cmd to run compiled executable:
```
export LSAN_OPTIONS=reports_objects=1:verbosity=0:detect_leaks=1
./a.out
```

The above compiled executable run to complete successfully, however, memory leak detected. See error log below:
```
2019-11-10 14:37:49.126448: I tensorflow/cc/saved_model/reader.cc:31] Reading SavedModel   from: /home/UNIXHOME/yli/repo/ccsqv/models/tf_multinom_2019_10_28
2019-11-10 14:37:49.131680: I tensorflow/cc/saved_model/reader.cc:54] Reading meta graph   with tags { serve }
2019-11-10 14:37:49.131734: I tensorflow/cc/saved_model/loader.cc:264] Reading SavedModel  debug info (if present) from: /home/UNIXHOME/yli/repo/ccsqv/models/tf_multinom_2019_10_28
2019-11-10 14:37:49.131881: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU  supports instructions that this TensorFlow binary was not compiled to use: SSE3 SSE4.1     SSE4.2
2019-11-10 14:37:49.205028: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU  Frequency: 2925790000 Hz
2019-11-10 14:37:49.214318: I tensorflow/compiler/xla/service/service.cc:168] XLA service  0x614000006c40 initialized for platform Host (this does not guarantee that XLA will be     used). Devices:
2019-11-10 14:37:49.214379: I tensorflow/compiler/xla/service/service.cc:176]              StreamExecutor device (0): Host, Default Version
2019-11-10 14:37:49.261663: I tensorflow/cc/saved_model/loader.cc:203] Restoring           SavedModel bundle.
2019-11-10 14:37:49.365354: I tensorflow/cc/saved_model/loader.cc:152] Running             initialization op on SavedModel bundle at path: /home/UNIXHOME/yli/repo/ccsqv/models/      tf_multinom_2019_10_28
2019-11-10 14:37:49.386885: I tensorflow/cc/saved_model/loader.cc:333] SavedModel load for tags { serve }; Status: success: OK. Took 260467 microseconds.

=================================================================
==14480==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 24 byte(s) in 1 object(s) allocated from:
    #0 0x7f0928f47a69 in operator new(unsigned long) /build/gcc-tc-passfinal-targ2targ//src/gcc/libsanitizer/asan/asan_new_delete.cc:90
    #1 0x7f09212aca67 in tensorflow::RegisterXlaDeviceKernels(char const*, char const*) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_cc.so.2+0x233ea67)
    #2 0x7f092129e88a in tensorflow::XlaCpuDeviceFactory::CreateDevices(tensorflow::SessionOptions const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::    allocator<char> > const&, std::vector<std::unique_ptr<tensorflow::Device, std::default_delete<tensorflow::Device> >, std::allocator<std::unique_ptr<tensorflow::Device, std::default_delete<tensorflow::Device> > > >*) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_cc.so.2+0x233088a)
    #3 0x7f091e611fb8 in tensorflow::DeviceFactory::AddDevices(tensorflow::SessionOptions const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::             allocator<char> > const&, std::vector<std::unique_ptr<tensorflow::Device, std::default_delete<tensorflow::Device> >, std::allocator<std::unique_ptr<tensorflow::Device, std::default_delete<tensorflow::Device> > > >*) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_framework.so.2+0xef3fb8)
    #4 0x7f0927491ae2 in tensorflow::DirectSessionFactory::NewSession(tensorflow::SessionOptions const&, tensorflow::Session**) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/    libtensorflow_cc.so.2+0x8523ae2)
    #5 0x7f091e6921af in tensorflow::NewSession(tensorflow::SessionOptions const&, tensorflow::Session**) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_framework.  so.2+0xf741af)
    #6 0x7f091dee5d15 in tensorflow::LoadSavedModel(tensorflow::SessionOptions const&, tensorflow::RunOptions const&, std::__cxx11::basic_string<char, std::                 char_traits<char>, std::allocator<char> > const&, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::        __cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >  >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, tensorflow::SavedModelBundle*) (/home/UNIXHOME/yli/venv/tf-r2. 1-cpu/lib/libtensorflow_framework.so.2+0x7c7d15)
    #7 0x405a32 in main (/home/UNIXHOME/yli/repo/ccsqv/a.out+0x405a32)
    #8 0x7f091c361c04 in __libc_start_main (/lib64/libc.so.6+0x21c04)

Indirect leak of 4096 byte(s) in 1 object(s) allocated from:
    #0 0x7f0928f47a69 in operator new(unsigned long) /build/gcc-tc-passfinal-targ2targ//src/gcc/libsanitizer/asan/asan_new_delete.cc:90
    #1 0x7f09212ac3a8 in void std::vector<std::unique_ptr<tensorflow::kernel_factory::OpKernelRegistrar, std::default_delete<tensorflow::kernel_factory::OpKernelRegistrar>  >, std::allocator<std::unique_ptr<tensorflow::kernel_factory::OpKernelRegistrar, std::default_delete<tensorflow::kernel_factory::OpKernelRegistrar> > > >::                  _M_realloc_insert<tensorflow::kernel_factory::OpKernelRegistrar*>(__gnu_cxx::__normal_iterator<std::unique_ptr<tensorflow::kernel_factory::OpKernelRegistrar, std::          default_delete<tensorflow::kernel_factory::OpKernelRegistrar> >*, std::vector<std::unique_ptr<tensorflow::kernel_factory::OpKernelRegistrar, std::default_delete<tensorflow::kernel_factory::OpKernelRegistrar> >, std::allocator<std::unique_ptr<tensorflow::kernel_factory::OpKernelRegistrar, std::default_delete<tensorflow::kernel_factory::         OpKernelRegistrar> > > > >, tensorflow::kernel_factory::OpKernelRegistrar*&&) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_cc.so.2+0x233e3a8)
    #2 0x7f09212acc6e in tensorflow::RegisterXlaDeviceKernels(char const*, char const*) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_cc.so.2+0x233ec6e)
    #3 0x7f092129e88a in tensorflow::XlaCpuDeviceFactory::CreateDevices(tensorflow::SessionOptions const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::    allocator<char> > const&, std::vector<std::unique_ptr<tensorflow::Device, std::default_delete<tensorflow::Device> >, std::allocator<std::unique_ptr<tensorflow::Device, std::default_delete<tensorflow::Device> > > >*) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_cc.so.2+0x233088a)
    #4 0x7f091e611fb8 in tensorflow::DeviceFactory::AddDevices(tensorflow::SessionOptions const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::             allocator<char> > const&, std::vector<std::unique_ptr<tensorflow::Device, std::default_delete<tensorflow::Device> >, std::allocator<std::unique_ptr<tensorflow::Device, std::default_delete<tensorflow::Device> > > >*) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_framework.so.2+0xef3fb8)
    #5 0x7f0927491ae2 in tensorflow::DirectSessionFactory::NewSession(tensorflow::SessionOptions const&, tensorflow::Session**) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/    libtensorflow_cc.so.2+0x8523ae2)
    #6 0x7f091e6921af in tensorflow::NewSession(tensorflow::SessionOptions const&, tensorflow::Session**) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_framework.  so.2+0xf741af)                                                                                                                                                                   #7 0x7f091dee5d15 in tensorflow::LoadSavedModel(tensorflow::SessionOptions const&, tensorflow::RunOptions const&, std::__cxx11::basic_string<char, std::                 char_traits<char>, std::allocator<char> > const&, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::        __cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >  >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, tensorflow::SavedModelBundle*) (/home/UNIXHOME/yli/venv/tf-r2. 1-cpu/lib/libtensorflow_framework.so.2+0x7c7d15)
    #8 0x405a32 in main (/home/UNIXHOME/yli/repo/ccsqv/a.out+0x405a32)
    #9 0x7f091c361c04 in __libc_start_main (/lib64/libc.so.6+0x21c04)

Indirect leak of 376 byte(s) in 376 object(s) allocated from:
    #0 0x7f0928f47a69 in operator new(unsigned long) /build/gcc-tc-passfinal-targ2targ//src/gcc/libsanitizer/asan/asan_new_delete.cc:90
    #1 0x7f09212acb47 in tensorflow::RegisterXlaDeviceKernels(char const*, char const*) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_cc.so.2+0x233eb47)
    #2 0x7f092129e88a in tensorflow::XlaCpuDeviceFactory::CreateDevices(tensorflow::SessionOptions const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::    allocator<char> > const&, std::vector<std::unique_ptr<tensorflow::Device, std::default_delete<tensorflow::Device> >, std::allocator<std::unique_ptr<tensorflow::Device, std::default_delete<tensorflow::Device> > > >*) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_cc.so.2+0x233088a)
    #3 0x7f091e611fb8 in tensorflow::DeviceFactory::AddDevices(tensorflow::SessionOptions const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::             allocator<char> > const&, std::vector<std::unique_ptr<tensorflow::Device, std::default_delete<tensorflow::Device> >, std::allocator<std::unique_ptr<tensorflow::Device, std::default_delete<tensorflow::Device> > > >*) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_framework.so.2+0xef3fb8)
    #4 0x7f0927491ae2 in tensorflow::DirectSessionFactory::NewSession(tensorflow::SessionOptions const&, tensorflow::Session**) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/    libtensorflow_cc.so.2+0x8523ae2)
    #5 0x7f091e6921af in tensorflow::NewSession(tensorflow::SessionOptions const&, tensorflow::Session**) (/home/UNIXHOME/yli/venv/tf-r2.1-cpu/lib/libtensorflow_framework.  so.2+0xf741af)
    #6 0x7f091dee5d15 in tensorflow::LoadSavedModel(tensorflow::SessionOptions const&, tensorflow::RunOptions const&, std::__cxx11::basic_string<char, std::                 char_traits<char>, std::allocator<char> > const&, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::        __cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >  >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, tensorflow::SavedModelBundle*) (/home/UNIXHOME/yli/venv/tf-r2. 1-cpu/lib/libtensorflow_framework.so.2+0x7c7d15)
    #7 0x405a32 in main (/home/UNIXHOME/yli/repo/ccsqv/a.out+0x405a32)
    #8 0x7f091c361c04 in __libc_start_main (/lib64/libc.so.6+0x21c04)

SUMMARY: AddressSanitizer: 4496 byte(s) leaked in 378 allocation(s).
```
The above leak was seen in both release `v2.0.0` and branch `r2.1`. 

This may be able to explain memory leak found in python LoadSavedModel.

To prevent further memory leakage, creating a nightly build with ASAN, USAN, LeakSanitizer and cover the core APIs may be a good idea to detect them early.

Would you please take a look at this issue? Thank you!

**Comment by ylipacbio** [comment]
To avoid running into memory leakage caused by XLA support, we recompiled tensorflow library without XLA support e.g., `build:xla --define with_xla_support=false`, and confirmed that Leak sanitizer check runs OK.

This was our work-around to this bug, not a FIX to this bug.

**Comment by sanjoy** [comment]
I don't think this is the root cause.  Since we have `static XlaDeviceOpRegistrations* registrations = ...` the allocation should happen only once and not cause a gradual memory leak.

Moreover, we need the `XlaDeviceOpRegistrations` to be alive for the entire process lifetime.

**Comment by ylipacbio** [comment]
Thanks for the reply and pointing out that `static XlaDeviceOpRegistrations* registrations = ` looks fine.

Please let me know once the root cause is found and fixed. Thanks!

**Comment by yhliang2018** [comment]
@julyrashchenko I've tried the model in the example with tf-nightly, but don't see a huge increase of the memory usage. Maybe the model is not large enough? Can you try tf-nightly for your real training job? We have fixed several memory leaking issues recently, and this problem may be potentially fixed already.

**Comment by jvishnuvardhan** [comment]
@julyrashchenko I just your code and I cannot reproduce the issue. I compared results from my earlier run (posted above) with the current run with `tf-nightly-gpu`. [Here](https://colab.sandbox.google.com/gist/jvishnuvardhan/997e4ce4c826fc15521f6c4eec4234e9/untitled670.ipynb) is the gist for your reference. 

Earlier, it used to leak ~100MB per 100 iterations. Currently it is way lower (~2 MB to 4 MB) per 100 iterations.

I am closing this issue as it was resolved in the `tf-nightly`. Please feel free to reopen if the issue persists again. Thanks!

**Comment by julyrashchenko** [comment]
> @julyrashchenko I've tried the model in the example with tf-nightly, but don't see a huge increase of the memory usage. Maybe the model is not large enough? Can you try tf-nightly for your real training job? We have fixed several memory leaking issues recently, and this problem may be potentially fixed already.

> @julyrashchenko I just your code and I cannot reproduce the issue. I compared results from my earlier run (posted above) with the current run with `tf-nightly-gpu`. [Here](https://colab.sandbox.google.com/gist/jvishnuvardhan/997e4ce4c826fc15521f6c4eec4234e9/untitled670.ipynb) is the gist for your reference.
> 
> Earlier, it used to leak ~100MB per 100 iterations. Currently it is way lower (~2 MB to 4 MB) per 100 iterations.
> 
> I am closing this issue as it was resolved in the `tf-nightly`. Please feel free to reopen if the issue persists again. Thanks!

@yhliang2018 @jvishnuvardhan Yes, I confirm that memory leak dissapeared with such setting when running on Google colab. But unfortunately I still suffer from it when running the same code on my machine, although much less with `tf-nightly` than before (about 100 Mb per 100 iterations now). Maybe, someone also encounters this problem? Or there is something I could not have taken into account?

**Comment by jvishnuvardhan** [comment]
@julyrashchenko What is the TF version you installed? If you have installed latest `tf-nightly` as in my google colab, you should be noticing similar performance. How much leak you are noticing now on your local machine? Thanks!

**Comment by julyrashchenko** [comment]
@jvishnuvardhan I installed with `pip install tf-nightly-gpu`, `tf.__version__` shows `2.1.0-dev20191126`.  Besides I tested it on one other machine and got the same results. With this code, memory leakage varies up to 100 Mb per 100 iterations. I was waiting till the end, but memory was never freed, it just decreased to zero, and then the program stopped. I attach the logs from my terminal.

```
$ python3 test_mem_leak.py
2019-11-27 10:09:39.775448: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
tfversion =  2.1.0-dev20191126
2019-11-27 10:09:40.578410: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2019-11-27 10:09:40.619368: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2019-11-27 10:09:40.619875: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1555] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: GeForce GTX 1050 Ti computeCapability: 6.1
coreClock: 1.392GHz coreCount: 6 deviceMemorySize: 3.94GiB deviceMemoryBandwidth: 104.43GiB/s
2019-11-27 10:09:40.619893: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
2019-11-27 10:09:40.666372: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2019-11-27 10:09:40.695804: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcufft.so.10
2019-11-27 10:09:40.703159: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcurand.so.10
2019-11-27 10:09:40.754736: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusolver.so.10
2019-11-27 10:09:40.763086: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusparse.so.10
2019-11-27 10:09:40.849905: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2019-11-27 10:09:40.850212: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2019-11-27 10:09:40.851218: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2019-11-27 10:09:40.852249: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1697] Adding visible gpu devices: 0
2019-11-27 10:09:40.852961: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-11-27 10:09:40.896006: I tensorflow/core/platform/profile_utils/cpu_utils.cc:101] CPU Frequency: 2808000000 Hz
2019-11-27 10:09:40.896977: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x344f4a0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2019-11-27 10:09:40.897046: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2019-11-27 10:09:40.976757: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2019-11-27 10:09:40.977051: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x411bee0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2019-11-27 10:09:40.977068: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): GeForce GTX 1050 Ti, Compute Capability 6.1
2019-11-27 10:09:40.977163: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2019-11-27 10:09:40.977366: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1555] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: GeForce GTX 1050 Ti computeCapability: 6.1
coreClock: 1.392GHz coreCount: 6 deviceMemorySize: 3.94GiB deviceMemoryBandwidth: 104.43GiB/s
2019-11-27 10:09:40.977384: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
2019-11-27 10:09:40.977395: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2019-11-27 10:09:40.977403: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcufft.so.10
2019-11-27 10:09:40.977412: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcurand.so.10
2019-11-27 10:09:40.977421: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusolver.so.10
2019-11-27 10:09:40.977428: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusparse.so.10
2019-11-27 10:09:40.977435: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2019-11-27 10:09:40.977469: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2019-11-27 10:09:40.977677: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2019-11-27 10:09:40.977859: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1697] Adding visible gpu devices: 0
2019-11-27 10:09:41.359293: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1096] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-11-27 10:09:41.359323: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102]      0 
2019-11-27 10:09:41.359330: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] 0:   N 
2019-11-27 10:09:41.359455: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2019-11-27 10:09:41.359711: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2019-11-27 10:09:41.359913: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1241] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 3048 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1050 Ti, pci bus id: 0000:01:00.0, compute capability: 6.1)
2019-11-27 10:09:41.562306: W tensorflow/python/util/util.cc:319] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
0 : free memory 9856.7734375 Mb
100 : free memory 9757.01171875 Mb
200 : free memory 9667.04296875 Mb
300 : free memory 9580.7421875 Mb
400 : free memory 9500.4921875 Mb
500 : free memory 9397.984375 Mb
600 : free memory 9156.78515625 Mb
700 : free memory 9087.34765625 Mb
800 : free memory 9027.28125 Mb
900 : free memory 9001.51171875 Mb
1000 : free memory 8924.78125 Mb
1100 : free memory 8836.68359375 Mb
1200 : free memory 8741.44140625 Mb
1300 : free memory 8663.7421875 Mb
1400 : free memory 8587.09375 Mb
1500 : free memory 8504.03125 Mb
```

**Comment by jvishnuvardhan** [comment]
@julyrashchenko Not sure what is the root-cause. In colab it is working well as I mentioned in my last comment. Thanks!

**Comment by vahdat-ab** [comment]
I am also getting a memory leak in my setup, which does active learning. It required lots of saving and loading. I am running my processes inside dockers and after a while dockers shut down my processes because they acquire memory and never release it.
I am running 2.0 and will try ```tf-nightly-gpu```.

**Comment by julyrashchenko** [comment]
I've managed to find a workaround for this issue which satisfied me. This is to save a model with `h5` format. With this approach there is no memory leak even with tf-2.0, and besides there are other benefits:
1. models are saved and loaded much faster (in my case ~4 times faster);
2. loaded models take much less memory (10 times less);
3. prediction also goes faster (~5 times).

And it didn't require many modifications to apply this approach,  essentially I had to implement method `get_config` for my custom layers.

Also if you need immediate memory deallocation, you can use `tf.keras.backend.clear_session()` and `gc.collect()`, they work fine now.

Here is modified code from my [gist](https://colab.research.google.com/drive/1If-E-UR23-ytIIIYctXScoJOMxZp42Xc):
```
import tensorflow as tf
import psutil
import gc

input = tf.keras.Input((1, 1024))
dense1 = tf.keras.layers.Dense(1024)(input)
dense2 = tf.keras.layers.Dense(1024)(dense1)
dense2 = tf.keras.layers.BatchNormalization()(dense2)
dense2 = tf.keras.layers.LeakyReLU()(dense2)
output = tf.keras.layers.Dense(1)(dense2)
model = tf.keras.Model(inputs=[input], outputs=[output])

def func():
  export_path = "temp_export.h5"
  model.save(export_path)
  tf.keras.models.load_model(export_path)
  tf.keras.backend.clear_session()

for i in range(1000000):
    func()
    if i % 100 == 0:
        print(i, ": free memory", psutil.virtual_memory().available / (1024.0 ** 2), "Mb") 
    gc.collect()
```

**Comment by jvishnuvardhan** [comment]
@julyrashchenko Thanks for the update.
Another important thing as mentioned by @k-w-w in another [post](https://github.com/tensorflow/tensorflow/issues/28923#issuecomment-521027273)

> The best practices can be summed up to:
If you are planning to load the model back into Keras, use the Keras APIs for saving and loading. If you're not planning on using the Keras APIs to further train the model, then tf.saved_model.save and tf.saved_model.load is sufficient. 

I am closing this issue as it was resolved. Please feel free to reopen if the issue persists again. Thanks!

**Comment by SSaishruthi** [comment]
Similar issue #40171

**Comment by mayerjTNG** [comment]
Hi, the fix with `.h5` doesn't do the trick for me, i still get memory leaks with both `save` and `save_weights` on tensorflow==2.7.0:

```
import gc

import psutil
import tensorflow as tf
from tqdm import tqdm

input = tf.keras.Input((1, 1024))
dense1 = tf.keras.layers.Dense(1024)(input)
dense2 = tf.keras.layers.Dense(1024)(dense1)
dense2 = tf.keras.layers.BatchNormalization()(dense2)
dense2 = tf.keras.layers.LeakyReLU()(dense2)
output = tf.keras.layers.Dense(1)(dense2)
model = tf.keras.Model(inputs=[input], outputs=[output])


def func():
    export_path = "temp_export.h5"
    model.save(export_path)
    tf.keras.backend.clear_session()
    gc.collect()


first_memory_usage = psutil.virtual_memory().available
progress_bar = tqdm()
step = 0
while True:
    func()
    progress_bar.set_description(
        f"Already lost {(first_memory_usage - psutil.virtual_memory().available) / (1024 ** 2):.3f} MB from tf.keras.Model.save"
    )
    step += 1
    progress_bar.update(step)
    ```
   ```
import gc

import psutil
import tensorflow as tf
from tqdm import tqdm

if __name__ == "__main__":

    input = tf.keras.Input((1, 1024))
    dense1 = tf.keras.layers.Dense(1024)(input)
    dense2 = tf.keras.layers.Dense(1024)(dense1)
    dense2 = tf.keras.layers.BatchNormalization()(dense2)
    dense2 = tf.keras.layers.LeakyReLU()(dense2)
    output = tf.keras.layers.Dense(1)(dense2)
    model = tf.keras.Model(inputs=[input], outputs=[output])

    first_memory_usage = psutil.virtual_memory().available
    progress_bar = tqdm()
    step = 0
    while True:
        model.save_weights("test.h5")

        tf.keras.backend.clear_session()
        gc.collect()
        progress_bar.set_description(
            f"Already lost {(first_memory_usage - psutil.virtual_memory().available) / (1024 ** 2):.3f} MB from tf.keras.Model.save_weights"
        )
        step += 1
        progress_bar.update(step)
           ```

**Comment by jvishnuvardhan** [comment]
@mayerjTNG As the closed issue was an old issue, Please open a new issue with the above standalone code to reproduce the issue. Thanks!
